File: packages/cli/hieudoanm/go/src/cmd/clipboard.go
package cmd

import (
	"database/sql"
	"fmt"
	"log"
	"time"

	"github.com/atotto/clipboard"
	_ "github.com/mattn/go-sqlite3"
	"github.com/spf13/cobra"
)

var clipboardCmd = &cobra.Command{
	Use:   "clipboard",
	Short: "Listen to clipboard changes and store them",
	Run: func(cmd *cobra.Command, args []string) {
		runClipboardWatcher()
	},
}

func init() {
	rootCmd.AddCommand(clipboardCmd)
}

// ── Core logic ───────────────────────────────────────────────

func runClipboardWatcher() {
	db, err := sql.Open("sqlite3", "~/.hieudoanm/hieudoanm.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	createTable(db)

	fmt.Println("📋 Clipboard watcher started... (Ctrl+C to stop)")

	var lastText string

	for {
		text, err := clipboard.ReadAll()
		if err != nil {
			time.Sleep(1 * time.Second)
			continue
		}

		// skip empty or same
		if text == "" || text == lastText {
			time.Sleep(500 * time.Millisecond)
			continue
		}

		err = insertClip(db, text)
		if err != nil {
			log.Println("insert error:", err)
		} else {
			fmt.Println("Saved:", preview(text))
		}

		lastText = text
		time.Sleep(500 * time.Millisecond)
	}
}

// ── DB ──────────────────────────────────────────────────────

func createTable(db *sql.DB) {
	query := `
	CREATE TABLE IF NOT EXISTS clips (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		content TEXT NOT NULL,
		created_at DATETIME DEFAULT CURRENT_TIMESTAMP
	);`
	_, err := db.Exec(query)
	if err != nil {
		log.Fatal(err)
	}
}

func insertClip(db *sql.DB, content string) error {
	_, err := db.Exec(
		`INSERT INTO clips (content) VALUES (?)`,
		content,
	)
	return err
}

// ── Utils ───────────────────────────────────────────────────

func preview(text string) string {
	if len(text) > 40 {
		return text[:40] + "..."
	}
	return text
}

File: packages/cli/hieudoanm/clap.rs/src/cmd/clipboard.rs
use rusqlite::Connection;
use std::path::PathBuf;
use std::time::Duration;

fn db_path() -> anyhow::Result<PathBuf> {
    let home = dirs::home_dir().ok_or_else(|| anyhow::anyhow!("cannot find home dir"))?;
    let dir = home.join(".hieudoanm");
    std::fs::create_dir_all(&dir)?;
    Ok(dir.join("hieudoanm.db"))
}

fn create_table(conn: &Connection) -> anyhow::Result<()> {
    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS clips (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            content TEXT NOT NULL,
            created_at DATETIME DEFAULT CURRENT_TIMESTAMP
        );",
    )?;
    Ok(())
}

fn insert_clip(conn: &Connection, content: &str) -> anyhow::Result<()> {
    conn.execute(
        "INSERT INTO clips (content) VALUES (?1)",
        rusqlite::params![content],
    )?;
    Ok(())
}

fn preview(text: &str) -> String {
    if text.len() > 40 {
        format!("{}...", &text[..40])
    } else {
        text.to_string()
    }
}

#[cfg(target_os = "macos")]
fn read_clipboard() -> anyhow::Result<String> {
    let output = std::process::Command::new("pbpaste")
        .output()
        .map_err(|e| anyhow::anyhow!("pbpaste failed: {e}"))?;
    let text = String::from_utf8_lossy(&output.stdout).to_string();
    Ok(text.trim().to_string())
}

#[cfg(target_os = "linux")]
fn read_clipboard() -> anyhow::Result<String> {
    let output = std::process::Command::new("xclip")
        .args(["-selection", "clipboard", "-o"])
        .output()
        .map_err(|e| anyhow::anyhow!("xclip failed: {e}"))?;
    let text = String::from_utf8_lossy(&output.stdout).to_string();
    Ok(text.trim().to_string())
}

#[cfg(not(any(target_os = "macos", target_os = "linux")))]
fn read_clipboard() -> anyhow::Result<String> {
    anyhow::bail!("clipboard not supported on this platform");
}

pub fn command() -> clap::Command {
    clap::Command::new("clipboard").about("Listen to clipboard changes and store them")
}

pub fn run(_matches: &clap::ArgMatches) -> anyhow::Result<()> {
    let path = db_path()?;
    let conn = Connection::open(&path)?;
    create_table(&conn)?;

    println!("Clipboard watcher started... (Ctrl+C to stop)");

    let mut last_text = String::new();

    loop {
        match read_clipboard() {
            Ok(text) if !text.is_empty() && text != last_text => {
                if let Err(e) = insert_clip(&conn, &text) {
                    eprintln!("insert error: {e}");
                } else {
                    println!("Saved: {}", preview(&text));
                }
                last_text = text;
            }
            _ => {}
        }
        std::thread::sleep(Duration::from_millis(500));
    }
}
